sn_node/tests/spend_simulation.rs

```diff
@@ -0,0 +1,697 @@
+        .await?;
+
+    for (id, address) in state.main_pubkeys.iter() {
+        let mut wallet = get_wallet(state.all_wallets.get(id).expect("Id should be present"));
+        wallet.deposit_and_store_to_disk(&transfer.cash_notes_for_recipient)?;
+        trace!(
+            "TestWallet {id} with main_pubkey: {address:?} has balance: {}",
+            wallet.balance()
+        );
+        assert_eq!(wallet.balance(), amount);
+
+        let (available_cash_notes, _lock) = wallet.available_cash_notes()?;
+
+        for (cashnote, _) in available_cash_notes {
+            state.cashnote_tracker.insert(
+                cashnote.unique_pubkey,
+                (SpendStatus::Utxo, cashnote.clone()),
+            );
+            match state.cashnotes_per_wallet.entry(*id) {
+                Entry::Vacant(entry) => {
+                    let _ = entry.insert(vec![cashnote.unique_pubkey]);
+                }
+                Entry::Occupied(entry) => entry.into_mut().push(cashnote.unique_pubkey),
+            }
+        }
+    }
+
+    Ok((client, state))
+}
+
+/// Returns random recipients to send tokens to.
+/// Random recipient of random lengths are chosen.
+fn get_recipients(our_id: usize, state: &State) -> Vec<MainPubkey> {
+    let mut recipients = Vec::new();
+
+    let mut random_number = our_id;
+    while random_number != our_id {
+        random_number = rand::thread_rng().gen_range(0..state.main_pubkeys.len());
+    }
+    recipients.push(state.main_pubkeys[&random_number]);
+
+    while random_number % 4 != 0 {
+        random_number = rand::thread_rng().gen_range(0..state.main_pubkeys.len());
+        if random_number != our_id {
+            recipients.push(state.main_pubkeys[&random_number]);
+        }
+    }
+
+    info!("TestWallet {our_id} the recipients for send are: {recipients:?}");
+    recipients
+}
+
+fn get_tx_to_attack(our_id: usize, state: &State) -> Result<Option<Transaction>> {
+    let mut rng = rand::thread_rng();
+    let Some(our_transactions) = state.outbound_transactions_per_wallet.get(&our_id) else {
+        info!("TestWallet {our_id} has no outbound transactions yet. Skipping attack");
+        return Ok(None);
+    };
+
+    if our_transactions.is_empty() {
+        info!("TestWallet {our_id} has no outbound transactions yet. Skipping attack");
+        return Ok(None);
+    }
+
+    let poisonable_tx = find_all_poisonable_spends(our_transactions, state)?;
+    if !poisonable_tx.is_empty() {
+        let random_tx = poisonable_tx
+            .into_iter()
+            .choose(&mut rng)
+            .ok_or_eyre("Cannot choose a random tx")?;
+
+        info!(
+            "TestWallet {our_id}. Poisoning transaction {:?}",
+            random_tx.hash()
+        );
+
+        return Ok(Some(random_tx.clone()));
+    }
+    Ok(None)
+}
+
+/// A spend / transaction is poisonable if all of its outputs are already spent.
+fn find_all_poisonable_spends<'a>(
```

Comment:
Actually, only one spent output is enough!

Reply:
If an output is unspent, then that output cannot be spent right. 
I wanted to address `poison`= even after a double spend, the entire branch is valid.